test/e2e/gardener/shoot/internal/inclusterclient/inclusterclient.go

```diff
@@ -0,0 +1,281 @@
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	imagevectorutils "github.com/gardener/gardener/pkg/utils/imagevector"
+	"github.com/gardener/gardener/test/framework"
+)
+
+const (
+	name      = "e2e-test-in-cluster-client"
+	namespace = metav1.NamespaceDefault
+
+	podNameDirect                 = name + "-direct"
+	podNameAPIServerProxyIP       = name + "-apiserver-proxy-ip"
+	podNameAPIServerProxyHostname = name + "-apiserver-proxy-hostname"
+	containerName                 = "kubectl"
+)
+
+var labels = map[string]string{"e2e-test": "in-cluster-client"}
+
+// VerifyInClusterAccessToAPIServer verifies access to the shoot API server from in-cluster clients.
+// It verifies the following paths:
+// - direct path using the KUBERNETES_SERVICE_HOST env var injected by gardener
+// - via the API server proxy's clusterIP
+// - via the API server proxy's hostname kubernetes.default.svc
+// For this, it deploys pods with the hyperkube image that contains the kubectl binary serving as a test client.
+// See docs/usage/networking/shoot_kubernetes_service_host_injection.md and docs/proposals/08-shoot-apiserver-via-sni.md
+func VerifyInClusterAccessToAPIServer(parentCtx context.Context, f *framework.ShootFramework) {
+	ctx, cancel := context.WithTimeout(parentCtx, 2*time.Minute)
+	defer cancel()
+
+	objects := getObjects(f.Shoot.Spec.Kubernetes.Version)
+	prepareObjects(ctx, f.ShootClient.Client(), objects...)
+
+	By("Verify in-cluster access to API server via direct path")
+	// this pod connects to the API server directly, i.e., uses the KUBERNETES_SERVICE_HOST env var injected by gardener
+	verifyAccessFromPod(ctx, f, podNameDirect, getInternalAPIServerAddress(f.Shoot))
+
+	By("Verify in-cluster access to API server via API server proxy IP")
+	// this pod connects via the API server proxy using the KUBERNETES_SERVICE_HOST env var injected by kubelet, i.e.,
+	// via the clusterIP of kubernetes.default.svc
+	verifyAccessFromPod(ctx, f, podNameAPIServerProxyIP, getInClusterAPIServerAddress(ctx, f.ShootClient.Client()))
+
+	By("Verify in-cluster access to API server via API server proxy hostname")
+	// this pod connects via the API server proxy hostname, i.e., via kubernetes.default.svc
+	verifyAccessFromPod(ctx, f, podNameAPIServerProxyHostname, "https://kubernetes.default.svc.cluster.local:443")
+}
+
+func verifyAccessFromPod(ctx context.Context, f *framework.ShootFramework, podName, expectedAddress string) {
+	By("Verify we are using the expected path")
+	Eventually(
+		execute(ctx, f.ShootClient, podName, "/kubectl", "cluster-info"),
+	).Should(Say(
+		"Kubernetes control plane is running at %s", expectedAddress,
+	))
+
+	By("Verify a typical API request works")
+	Eventually(
+		execute(ctx, f.ShootClient, podName, "/kubectl", "get", "pod", podName),
+	).Should(Say(
+		podName,
+	))
```

Comment:
Should we check for more than the pod name? If the pod does not exist this would also succeed, but then again we want to test that we get a response from `kube-apiserver`. Hence, it may be sufficient.
WDYT?